Layout fixes, two-column landscape first page

direc85 opened the pull request:
Hi,

I added landscape orientation in FirstPage.qml; it now uses two columns instead of one, which makes the layout a lot nicer. ProgramDelegate.qml has layout fixes so that using different devices should work properly now.

direc85 committed:
Ignore *.spec.* files

direc85 committed:
Layout fixes for different font sizes

direc85 committed:
Support landscape orientations

direc85 committed:
Add thumbnail selection highlight

direc85 committed:
Enable two-column landscape layout

Inspired by /usr/share/jolla-gallery/GalleryGridPage.qml

mlehtima commented on qml/pages/FirstPage.qml:
```diff
@@ -5,26 +5,47 @@ import "../js/yleApi.js" as YleApi
 
-    // To enable PullDownMenu, place our content in a SilicaFlickable
-    SilicaFlickable {
+    onOrientationChanged: rowCount = Math.ceil(model.length / columnCount)
```
model is a variable of gridView so should be gridView.model here

direc85 replied:
Correct, I'll fix that.

mlehtima commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +62,91 @@ Page {
+            id: listItem
+            width: page.width / columnCount
+            property int baseHeight: gridView.cellHeight
+            height: baseHeight + (index === curreIndex ? offsetY : 0.0)
```
Typo curreIndex should be gridView.currentIndex

direc85 replied:
It actually should be global currIndex, as using gridView.currentView causes jerkiness when long-pressing the first element in portrait mode. But yes, it's a typo.

mlehtima commented on qml/pages/ProgramDelegate.qml:
```diff
@@ -55,6 +67,7 @@ ListItem {
             font.bold: true
             font.pixelSize: Theme.fontSizeExtraSmall
             truncationMode: TruncationMode.Fade
+            y: parent.y + Theme.paddingSmall
```
Maybe use topMargin in anchors below instead? Seemed work the same way and is a bit nicer way.

direc85 replied:
That seems nicer option indeed. I've had some issues with anchors before, so I went manually with the y value.

mlehtima commented on qml/pages/ProgramDelegate.qml:
```diff
@@ -26,15 +27,18 @@ ListItem {
-            height: parent.height
+            height: parent.height - 2*Theme.paddingSmall
             width: Math.ceil((height * 16) / 9)
+            y: parent.y + Theme.paddingSmall
```
Not needed as topMargin is already used below in anchors.

direc85 replied:
Correct; fixed.

direc85 replied:
That's correct, I'll fix it.

mlehtima reviewed (changes requested):
Some small comments, maybe also consider adding padding/margins around the content in first page, it looks a bit odd when in landscape mode the content has no spacing. Also in portrait mode there should be margins according to Sailfish OS style instructions.

mlehtima commented on qml/pages/ProgramDelegate.qml:
```diff
@@ -6,7 +6,8 @@ import "../js/yleApi.js" as YleApi
     id: listItem
     contentWidth: ListView.view.width
-    contentHeight: Theme.itemSizeLarge
+    //contentHeight: Theme.itemSizeLarge
```
Does this predefined size look bad on some device? Or are the changes only because of the text spacing? Also do not leave commented out code unless there is a very good reason.

direc85 replied:
It's because of the text sizing; the content is either not centered properly or overflows (at least when using a different system font size). I haven't found more elegant way to do it, so I used this method. The commented code is indeed an oversight, I'll clean it up.

mlehtima commented on qml/pages/ProgramDelegate.qml:
```diff
@@ -26,15 +27,18 @@ ListItem {
-        height: parent.height - 2 * Theme.paddingSmall
+        id: listContainer
+        //height: parent.height - 2 * Theme.paddingSmall
+        height: title.height + episode.height + time.height + 2*Theme.paddingSmall
```
It would make these more simple if you add 2 * Theme.paddingSmall to listItem contentHeight and then matching anchor margins to this listContainer and then remove the top margins from elsewhere.

direc85 replied:
Thanks, I'll have a look at simplifying the layout calculations again.

joonne commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +62,91 @@ Page {
-            anchors.fill: parent
-            header: PageHeader {
-                title: qsTr("Current Broadcasts")
+        onModelChanged: rowCount = Math.ceil(model.length / columnCount)
```
Should this also be `gridView.model` for clarity? Most likely works since the model is bound inside the event handler I think.

direc85 replied:
I don't think it's necessary; we are inside onModelChanged, which is inside the gridView, too. I would leave this as-is.

joonne commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +62,91 @@ Page {
+        cellWidth: page.width / columnCount
+        cellHeight: page.width / columnCount / 16.0 * 9.0 + Theme.fontSizeMedium + Theme.paddingMedium
+
+        // Always, always remember to include the header itemi height...
```
Fix typo or remove whole comment if not necessary or maybe open with the comment `why` this needs to be remembered? :)

direc85 replied:
That was really a note to self - I always, always forget that... Removing the comment.

joonne commented on qml/pages/ProgramDelegate.qml:
```diff
@@ -26,15 +27,18 @@ ListItem {
             id: img
             color: "black"
-            height: parent.height
+            height: parent.height - 2*Theme.paddingSmall
```
`2 * Theme.paddingSmall`

direc85 replied:
Fixed by resolving the previous comment.

joonne reviewed (commented):
Hey :wave:  Welcome + thanks for contributing :) I think @mlehtima addressed all the relevant stuff but added few comments nevertheless.

direc85 committed:
Corrections made as suggested by joonne and mlehtima

direc85 commented:
I made the suggested changes to the code. I don't really know (yet) how to properly handle reviewing code in GitHub, so I just updated the PR.

direc85 committed:
Use BusyIndicator and fallback Image

direc85 committed:
Add Refresh to pulley menu

direc85 committed:
Update translations

direc85 commented:
Oh dang it. I forgot that pushing commits to master updates pull request, too...

Well, I implemented BusyIndicators while images are loading, and displaying a theme-based placeholder image if the image can't be loaded. It's used both on FirstPage and on ProgramDelegate.

I also added refresh to the first pages pulley menu.

joonne commented:
Would it be feasible for you to make a feature branch out of this actual repo, rebase your changes on top of it and submit a PR from your feature branch against master? You can then decide if you'd like the additional features included in the PR or make a new one for them? I added you already as a collaborator. Not sure though what is the usual convention here, make a fork and submit PR:s from there? Or from the same repo via feature branches as a collaborator?

joonne commented on qml/pages/FirstPage.qml:
```diff
@@ -5,32 +5,61 @@ import "../js/yleApi.js" as YleApi
 
 Page {
     id: page
+    property real rowCount
```
Should these have some initial values?

joonne commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +70,96 @@ Page {
+
+        // Cell width and height depend on columnCount, which depends on orientation
+        cellWidth: page.width / columnCount
+        cellHeight: page.width / columnCount / 16.0 * 9.0 + Theme.fontSizeMedium + Theme.paddingMedium
```
Where do these `Theme.fontSizeMedium` & `Theme.paddingMedium` come from?

joonne replied:
also: maybe add the aspect ratio as `readonly property ASPECT_RATIO: 16*9`

I suppose that is what it does?

mlehtima commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +70,96 @@ Page {
+                id: contextMenu
+
+                // Using offsetY the gridView size changes dynamically. Magic!
+                onHeightChanged: offsetY = height
```
I think this might be causing an issue I observe. If I long press the first item in list to show context menu and then try to click one of the items in the context menu the context menu offset changes before the click is registered and it will instead click the image and video will start to play instead of showing the new page based on menu click.

mlehtima commented on qml/pages/FirstPage.qml:
```diff
@@ -41,67 +70,96 @@ Page {
+                }
+            }
+
+            Label {
```
I would like to see some margin around the title, now it's very close to the image. At least some margin before the next image would be nice, I think I also mentioned earlier that margin on both left and right sides the image and title would make the first page follow Sailfish OS app guidelines.

direc85 commented:
I'm closing this and creating new brach(es) for the improvements, as discussed earlier.
